src/IO/RemoteReadBufferCache.h

```diff
@@ -0,0 +1,211 @@
+#pragma once
+#include <mutex>
+#include <list>
+#include <set>
+#include <map>
+#include <memory>
+#include <Poco/FileStream.h>
+#include <Poco/Logger.h>
+#include <Common/ThreadPool.h>
+#include <IO/ReadBuffer.h>
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/SeekableReadBuffer.h>
+#include <condition_variable>
+
+namespace DB
+{
+/**
+ *
+ */
+class RemoteCacheController
+{
+public:
+    RemoteCacheController(
+        const std::string & schema_,
+        const std::string & cluster_,
+        const std::string & path_,
+        UInt64 mod_ts,
+        const std::string & local_path_,
+        std::shared_ptr<ReadBuffer> readbuffer_,
+        std::function<void(RemoteCacheController *)> const & finish_callback);
+    ~RemoteCacheController(); // the local files will be deleted in descontructor
+
+    // recover from local disk
+    static std::shared_ptr<RemoteCacheController>
+    recover(const std::string & local_path, std::function<void(RemoteCacheController *)> const & finish_callback);
+
+    /**
+     * called by LocalCachedFileReader, must be used in pair
+     * local_path will be empty if the file has not been downloaded
+     */
+    FILE * allocFile(std::string * local_path);
+    void deallocFile(FILE * fs_);
+
+    /**
+     * when allocFile be called, count++. deallocFile be called, count--.
+     * the local file could be deleted only count==0
+     */
+    inline bool closable()
+    {
+        std::lock_guard lock{mutex};
+        return opened_file_streams.empty() && remote_readbuffer == nullptr;
+    }
+    void close();
+
+    /**
+     * called in LocalCachedFileReader read(), the reading process would be blocked until
+     * enough data be downloaded.
+     * If the file has finished download, the process would be block
+     * returns
+     *  - 0: has more data to read
+     *  - -1: has reach eof
+     */
+    int waitMoreData(size_t start_offset_, size_t end_offset_);
+
+    inline size_t size() const { return current_offset; }
+
+    inline const std::string & getLocalPath() { return local_path; }
+    inline const std::string & getRemotePath() { return remote_path; }
+
+    inline UInt64 getLastModTS() const { return last_mod_ts; }
+    inline void markInvalid()
+    {
+        std::lock_guard lock(mutex);
+        valid = false;
+    }
+    inline bool isValid()
+    {
+        std::lock_guard lock(mutex);
+        return valid;
+    }
+
+private:
+    // flush file and meta info into disk
+    void flush();
+
+    void backgroupDownload(std::function<void(RemoteCacheController *)> const & finish_callback);
+
+    std::mutex mutex;
+    std::condition_variable more_data_signal;
+    ThreadPool * download_thread;
+
+    std::set<FILE *> opened_file_streams;
```

Comment:
May be use a set of `ReadBufferFromFile` instead?

Reply:
I think that to use `ReadBufferFromFile` is more complate

Reply:
why? I do not think that it is worse.